<!DOCTYPE html>
<html data-markdown-preview-plus-context="html-export">
  <head>
    <meta charset="utf-8" />
    <title>jamboree_steps.md</title>
    <style>.emoji {
  max-width: 1em !important;
}
del {
  text-decoration: none;
  position: relative;
}
del::after {
  border-bottom: 1px solid black;
  content: '';
  left: 0;
  position: absolute;
  right: 0;
  top: 50%;
}
ul.contains-task-list li.task-list-item {
  position: relative;
  list-style-type: none;
}
ul.contains-task-list li.task-list-item input.task-list-item-checkbox {
  position: absolute;
  transform: translateX(-100%);
  width: 26px;
}
span.critic.comment {
  position: relative;
}
span.critic.comment::before {
  content: '\1f4ac';
  position: initial;
}
span.critic.comment > span {
  display: none;
}
span.critic.comment:hover > span {
  display: initial;
  position: absolute;
  top: 100%;
  left: 0;
  border: 1px solid;
  border-radius: 5px;
  max-height: 4em;
  overflow: auto;
}
span.critic.comment:focus > span {
  display: initial;
  text-decoration: underline;
  position: initial;
  top: auto;
  left: auto;
  border: initial;
  border-radius: initial;
}
table {
  border-collapse: collapse;
  border-spacing: 0;
  background-color: transparent;
}

body {
  overflow: initial !important;
  overflow: hidden;
  font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif;
  line-height: 1.6;
  word-wrap: break-word;
  padding: 30px;
  font-size: 16px;
  color: #333;
  background-color: #fff;
}
body > *:first-child {
  margin-top: 0 !important;
}
body > *:last-child {
  margin-bottom: 0 !important;
}
body a:not([href]) {
  color: inherit;
  text-decoration: none;
}
body .absent {
  color: #c00;
}
body .anchor {
  position: absolute;
  top: 0;
  left: 0;
  display: block;
  padding-right: 6px;
  padding-left: 30px;
  margin-left: -30px;
}
body .anchor:focus {
  outline: none;
}
body h1,
body h2,
body h3,
body h4,
body h5,
body h6 {
  position: relative;
  margin-top: 1em;
  margin-bottom: 16px;
  font-weight: bold;
  line-height: 1.4;
}
body h1 .octicon-link,
body h2 .octicon-link,
body h3 .octicon-link,
body h4 .octicon-link,
body h5 .octicon-link,
body h6 .octicon-link {
  display: none;
  color: #000;
  vertical-align: middle;
}
body h1:hover .anchor,
body h2:hover .anchor,
body h3:hover .anchor,
body h4:hover .anchor,
body h5:hover .anchor,
body h6:hover .anchor {
  padding-left: 8px;
  margin-left: -30px;
  text-decoration: none;
}
body h1:hover .anchor .octicon-link,
body h2:hover .anchor .octicon-link,
body h3:hover .anchor .octicon-link,
body h4:hover .anchor .octicon-link,
body h5:hover .anchor .octicon-link,
body h6:hover .anchor .octicon-link {
  display: inline-block;
}
body h1 tt,
body h2 tt,
body h3 tt,
body h4 tt,
body h5 tt,
body h6 tt,
body h1 code,
body h2 code,
body h3 code,
body h4 code,
body h5 code,
body h6 code {
  font-size: inherit;
}
body h1 {
  padding-bottom: 0.3em;
  font-size: 2.25em;
  line-height: 1.2;
  border-bottom: 1px solid #eee;
}
body h1 .anchor {
  line-height: 1;
}
body h2 {
  padding-bottom: 0.3em;
  font-size: 1.75em;
  line-height: 1.225;
  border-bottom: 1px solid #eee;
}
body h2 .anchor {
  line-height: 1;
}
body h3 {
  font-size: 1.5em;
  line-height: 1.43;
}
body h3 .anchor {
  line-height: 1.2;
}
body h4 {
  font-size: 1.25em;
}
body h4 .anchor {
  line-height: 1.2;
}
body h5 {
  font-size: 1em;
}
body h5 .anchor {
  line-height: 1.1;
}
body h6 {
  font-size: 1em;
  color: #777;
}
body h6 .anchor {
  line-height: 1.1;
}
body p,
body blockquote,
body ul,
body ol,
body dl,
body table,
body pre {
  margin-top: 0;
  margin-bottom: 16px;
}
body hr {
  height: 4px;
  padding: 0;
  margin: 16px 0;
  background-color: #e7e7e7;
  border: 0 none;
}
body ul,
body ol {
  padding-left: 2em;
}
body ul.no-list,
body ol.no-list {
  padding: 0;
  list-style-type: none;
}
body ul ul,
body ul ol,
body ol ol,
body ol ul {
  margin-top: 0;
  margin-bottom: 0;
}
body li > p {
  margin-top: 16px;
}
body dl {
  padding: 0;
}
body dl dt {
  padding: 0;
  margin-top: 16px;
  font-size: 1em;
  font-style: italic;
  font-weight: bold;
}
body dl dd {
  padding: 0 16px;
  margin-bottom: 16px;
}
body blockquote {
  padding: 0 15px;
  color: #777;
  border-left: 4px solid #ddd;
}
body blockquote > :first-child {
  margin-top: 0;
}
body blockquote > :last-child {
  margin-bottom: 0;
}
body table {
  display: block;
  width: 100%;
  overflow: auto;
  word-break: normal;
  word-break: keep-all;
}
body table th {
  font-weight: bold;
}
body table th,
body table td {
  padding: 6px 13px;
  border: 1px solid #ddd;
}
body table tr {
  background-color: #fff;
  border-top: 1px solid #ccc;
}
body table tr:nth-child(2n) {
  background-color: #f8f8f8;
}
body img {
  max-width: 100%;
  -moz-box-sizing: border-box;
  box-sizing: border-box;
}
body .emoji {
  max-width: none;
}
body span.frame {
  display: block;
  overflow: hidden;
}
body span.frame > span {
  display: block;
  float: left;
  width: auto;
  padding: 7px;
  margin: 13px 0 0;
  overflow: hidden;
  border: 1px solid #ddd;
}
body span.frame span img {
  display: block;
  float: left;
}
body span.frame span span {
  display: block;
  padding: 5px 0 0;
  clear: both;
  color: #333;
}
body span.align-center {
  display: block;
  overflow: hidden;
  clear: both;
}
body span.align-center > span {
  display: block;
  margin: 13px auto 0;
  overflow: hidden;
  text-align: center;
}
body span.align-center span img {
  margin: 0 auto;
  text-align: center;
}
body span.align-right {
  display: block;
  overflow: hidden;
  clear: both;
}
body span.align-right > span {
  display: block;
  margin: 13px 0 0;
  overflow: hidden;
  text-align: right;
}
body span.align-right span img {
  margin: 0;
  text-align: right;
}
body span.float-left {
  display: block;
  float: left;
  margin-right: 13px;
  overflow: hidden;
}
body span.float-left span {
  margin: 13px 0 0;
}
body span.float-right {
  display: block;
  float: right;
  margin-left: 13px;
  overflow: hidden;
}
body span.float-right > span {
  display: block;
  margin: 13px auto 0;
  overflow: hidden;
  text-align: right;
}
body code,
body tt {
  padding: 0;
  padding-top: 0.2em;
  padding-bottom: 0.2em;
  margin: 0;
  font-size: 85%;
  background-color: rgba(0, 0, 0, 0.04);
  border-radius: 3px;
}
body code:before,
body tt:before,
body code:after,
body tt:after {
  letter-spacing: -0.2em;
  content: "\00a0";
}
body code br,
body tt br {
  display: none;
}
body del code {
  text-decoration: inherit;
}
body pre > code {
  padding: 0;
  margin: 0;
  font-size: 100%;
  word-break: normal;
  white-space: pre;
  background: transparent;
  border: 0;
}
body .highlight {
  margin-bottom: 16px;
}
body .highlight pre,
body pre {
  padding: 16px;
  overflow: auto;
  font-size: 85%;
  line-height: 1.45;
  background-color: #f7f7f7;
  border-radius: 3px;
}
body .highlight pre {
  margin-bottom: 0;
  word-break: normal;
}
body pre {
  word-wrap: normal;
}
body pre code,
body pre tt {
  display: inline;
  max-width: initial;
  padding: 0;
  margin: 0;
  overflow: initial;
  line-height: inherit;
  word-wrap: normal;
  background-color: transparent;
  border: 0;
}
body pre code:before,
body pre tt:before,
body pre code:after,
body pre tt:after {
  content: normal;
}
body kbd {
  display: inline-block;
  padding: 3px 5px;
  font-size: 11px;
  line-height: 10px;
  color: #555;
  vertical-align: middle;
  background-color: #fcfcfc;
  border: solid 1px #ccc;
  border-bottom-color: #bbb;
  border-radius: 3px;
  box-shadow: inset 0 -1px 0 #bbb;
}
span.critic.comment > span {
  background-color: #fff;
}
a {
  color: #337ab7;
}

.bracket-matcher .region {
  border-bottom: 1px dotted lime;
  position: absolute;
}
.line-number.bracket-matcher.bracket-matcher {
  color: #abb2bf;
  background-color: #3a3f4b;
}

.spell-check-misspelling .region {
  border-bottom: 2px dotted rgba(255, 51, 51, 0.75);
}
.spell-check-corrections {
  width: 25em !important;
}

pre.editor-colors {
  background-color: #282c34;
  color: #abb2bf;
}
pre.editor-colors .line.cursor-line {
  background-color: rgba(153, 187, 255, 0.04);
}
pre.editor-colors .invisible {
  color: #abb2bf;
}
pre.editor-colors .cursor {
  border-left: 2px solid #528bff;
}
pre.editor-colors .selection .region {
  background-color: #3e4451;
}
pre.editor-colors .bracket-matcher .region {
  border-bottom: 1px solid #528bff;
  box-sizing: border-box;
}
pre.editor-colors .invisible-character {
  color: rgba(171, 178, 191, 0.15);
}
pre.editor-colors .indent-guide {
  color: rgba(171, 178, 191, 0.15);
}
pre.editor-colors .wrap-guide {
  background-color: rgba(171, 178, 191, 0.15);
}
pre.editor-colors .find-result .region.region.region,
pre.editor-colors .current-result .region.region.region {
  border-radius: 2px;
  background-color: rgba(82, 139, 255, 0.24);
  transition: border-color 0.4s;
}
pre.editor-colors .find-result .region.region.region {
  border: 2px solid transparent;
}
pre.editor-colors .current-result .region.region.region {
  border: 2px solid #528bff;
  transition-duration: .1s;
}
pre.editor-colors .gutter .line-number {
  color: #636d83;
  -webkit-font-smoothing: antialiased;
}
pre.editor-colors .gutter .line-number.cursor-line {
  color: #abb2bf;
  background-color: #3a3f4b;
}
pre.editor-colors .gutter .line-number.cursor-line-no-selection {
  background-color: transparent;
}
pre.editor-colors .gutter .line-number .icon-right {
  color: #abb2bf;
}
pre.editor-colors .gutter:not(.git-diff-icon) .line-number.git-line-removed.git-line-removed::before {
  bottom: -3px;
}
pre.editor-colors .gutter:not(.git-diff-icon) .line-number.git-line-removed::after {
  content: "";
  position: absolute;
  left: 0px;
  bottom: 0px;
  width: 25px;
  border-bottom: 1px dotted rgba(224, 82, 82, 0.5);
  pointer-events: none;
}
pre.editor-colors .gutter .line-number.folded,
pre.editor-colors .gutter .line-number:after,
pre.editor-colors .fold-marker:after {
  color: #abb2bf;
}
.syntax--comment {
  color: #5c6370;
  font-style: italic;
}
.syntax--comment .syntax--markup.syntax--link {
  color: #5c6370;
}
.syntax--entity.syntax--name.syntax--type {
  color: #e5c07b;
}
.syntax--entity.syntax--other.syntax--inherited-class {
  color: #e5c07b;
}
.syntax--keyword {
  color: #c678dd;
}
.syntax--keyword.syntax--control {
  color: #c678dd;
}
.syntax--keyword.syntax--operator {
  color: #c678dd;
}
.syntax--keyword.syntax--other.syntax--special-method {
  color: #61afef;
}
.syntax--keyword.syntax--other.syntax--unit {
  color: #d19a66;
}
.syntax--storage {
  color: #c678dd;
}
.syntax--storage.syntax--type.syntax--annotation,
.syntax--storage.syntax--type.syntax--primitive {
  color: #c678dd;
}
.syntax--storage.syntax--modifier.syntax--package,
.syntax--storage.syntax--modifier.syntax--import {
  color: #abb2bf;
}
.syntax--constant {
  color: #d19a66;
}
.syntax--constant.syntax--variable {
  color: #d19a66;
}
.syntax--constant.syntax--character.syntax--escape {
  color: #56b6c2;
}
.syntax--constant.syntax--numeric {
  color: #d19a66;
}
.syntax--constant.syntax--other.syntax--color {
  color: #56b6c2;
}
.syntax--constant.syntax--other.syntax--symbol {
  color: #56b6c2;
}
.syntax--variable {
  color: #e06c75;
}
.syntax--variable.syntax--interpolation {
  color: #be5046;
}
.syntax--variable.syntax--parameter {
  color: #abb2bf;
}
.syntax--string {
  color: #98c379;
}
.syntax--string > .syntax--source,
.syntax--string .syntax--embedded {
  color: #abb2bf;
}
.syntax--string.syntax--regexp {
  color: #56b6c2;
}
.syntax--string.syntax--regexp .syntax--source.syntax--ruby.syntax--embedded {
  color: #e5c07b;
}
.syntax--string.syntax--other.syntax--link {
  color: #e06c75;
}
.syntax--punctuation.syntax--definition.syntax--comment {
  color: #5c6370;
}
.syntax--punctuation.syntax--definition.syntax--method-parameters,
.syntax--punctuation.syntax--definition.syntax--function-parameters,
.syntax--punctuation.syntax--definition.syntax--parameters,
.syntax--punctuation.syntax--definition.syntax--separator,
.syntax--punctuation.syntax--definition.syntax--seperator,
.syntax--punctuation.syntax--definition.syntax--array {
  color: #abb2bf;
}
.syntax--punctuation.syntax--definition.syntax--heading,
.syntax--punctuation.syntax--definition.syntax--identity {
  color: #61afef;
}
.syntax--punctuation.syntax--definition.syntax--bold {
  color: #e5c07b;
  font-weight: bold;
}
.syntax--punctuation.syntax--definition.syntax--italic {
  color: #c678dd;
  font-style: italic;
}
.syntax--punctuation.syntax--section.syntax--embedded {
  color: #be5046;
}
.syntax--punctuation.syntax--section.syntax--method,
.syntax--punctuation.syntax--section.syntax--class,
.syntax--punctuation.syntax--section.syntax--inner-class {
  color: #abb2bf;
}
.syntax--support.syntax--class {
  color: #e5c07b;
}
.syntax--support.syntax--type {
  color: #56b6c2;
}
.syntax--support.syntax--function {
  color: #56b6c2;
}
.syntax--support.syntax--function.syntax--any-method {
  color: #61afef;
}
.syntax--entity.syntax--name.syntax--function {
  color: #61afef;
}
.syntax--entity.syntax--name.syntax--class,
.syntax--entity.syntax--name.syntax--type.syntax--class {
  color: #e5c07b;
}
.syntax--entity.syntax--name.syntax--section {
  color: #61afef;
}
.syntax--entity.syntax--name.syntax--tag {
  color: #e06c75;
}
.syntax--entity.syntax--other.syntax--attribute-name {
  color: #d19a66;
}
.syntax--entity.syntax--other.syntax--attribute-name.syntax--id {
  color: #61afef;
}
.syntax--meta.syntax--class {
  color: #e5c07b;
}
.syntax--meta.syntax--class.syntax--body {
  color: #abb2bf;
}
.syntax--meta.syntax--method-call,
.syntax--meta.syntax--method {
  color: #abb2bf;
}
.syntax--meta.syntax--definition.syntax--variable {
  color: #e06c75;
}
.syntax--meta.syntax--link {
  color: #d19a66;
}
.syntax--meta.syntax--require {
  color: #61afef;
}
.syntax--meta.syntax--selector {
  color: #c678dd;
}
.syntax--meta.syntax--separator {
  color: #abb2bf;
}
.syntax--meta.syntax--tag {
  color: #abb2bf;
}
.syntax--underline {
  text-decoration: underline;
}
.syntax--none {
  color: #abb2bf;
}
.syntax--invalid.syntax--deprecated {
  color: #523d14 !important;
  background-color: #e0c285 !important;
}
.syntax--invalid.syntax--illegal {
  color: white !important;
  background-color: #e05252 !important;
}
.syntax--markup.syntax--bold {
  color: #d19a66;
  font-weight: bold;
}
.syntax--markup.syntax--changed {
  color: #c678dd;
}
.syntax--markup.syntax--deleted {
  color: #e06c75;
}
.syntax--markup.syntax--italic {
  color: #c678dd;
  font-style: italic;
}
.syntax--markup.syntax--heading {
  color: #e06c75;
}
.syntax--markup.syntax--heading .syntax--punctuation.syntax--definition.syntax--heading {
  color: #61afef;
}
.syntax--markup.syntax--link {
  color: #56b6c2;
}
.syntax--markup.syntax--inserted {
  color: #98c379;
}
.syntax--markup.syntax--quote {
  color: #d19a66;
}
.syntax--markup.syntax--raw {
  color: #98c379;
}
.syntax--source.syntax--cs .syntax--keyword.syntax--operator {
  color: #c678dd;
}
.syntax--source.syntax--css .syntax--property-name,
.syntax--source.syntax--css .syntax--property-value {
  color: #828997;
}
.syntax--source.syntax--css .syntax--property-name.syntax--support,
.syntax--source.syntax--css .syntax--property-value.syntax--support {
  color: #abb2bf;
}
.syntax--source.syntax--elixir .syntax--source.syntax--embedded.syntax--source {
  color: #abb2bf;
}
.syntax--source.syntax--elixir .syntax--constant.syntax--language,
.syntax--source.syntax--elixir .syntax--constant.syntax--numeric,
.syntax--source.syntax--elixir .syntax--constant.syntax--definition {
  color: #61afef;
}
.syntax--source.syntax--elixir .syntax--variable.syntax--definition,
.syntax--source.syntax--elixir .syntax--variable.syntax--anonymous {
  color: #c678dd;
}
.syntax--source.syntax--elixir .syntax--parameter.syntax--variable.syntax--function {
  color: #d19a66;
  font-style: italic;
}
.syntax--source.syntax--elixir .syntax--quoted {
  color: #98c379;
}
.syntax--source.syntax--elixir .syntax--keyword.syntax--special-method,
.syntax--source.syntax--elixir .syntax--embedded.syntax--section,
.syntax--source.syntax--elixir .syntax--embedded.syntax--source.syntax--empty {
  color: #e06c75;
}
.syntax--source.syntax--elixir .syntax--readwrite.syntax--module .syntax--punctuation {
  color: #e06c75;
}
.syntax--source.syntax--elixir .syntax--regexp.syntax--section,
.syntax--source.syntax--elixir .syntax--regexp.syntax--string {
  color: #be5046;
}
.syntax--source.syntax--elixir .syntax--separator,
.syntax--source.syntax--elixir .syntax--keyword.syntax--operator {
  color: #d19a66;
}
.syntax--source.syntax--elixir .syntax--variable.syntax--constant {
  color: #e5c07b;
}
.syntax--source.syntax--elixir .syntax--array,
.syntax--source.syntax--elixir .syntax--scope,
.syntax--source.syntax--elixir .syntax--section {
  color: #828997;
}
.syntax--source.syntax--gfm .syntax--markup {
  -webkit-font-smoothing: auto;
}
.syntax--source.syntax--gfm .syntax--link .syntax--entity {
  color: #61afef;
}
.syntax--source.syntax--ini .syntax--keyword.syntax--other.syntax--definition.syntax--ini {
  color: #e06c75;
}
.syntax--source.syntax--java .syntax--storage.syntax--modifier.syntax--import {
  color: #e5c07b;
}
.syntax--source.syntax--java .syntax--storage.syntax--type {
  color: #e5c07b;
}
.syntax--source.syntax--java .syntax--keyword.syntax--operator.syntax--instanceof {
  color: #c678dd;
}
.syntax--source.syntax--java-properties .syntax--meta.syntax--key-pair {
  color: #e06c75;
}
.syntax--source.syntax--java-properties .syntax--meta.syntax--key-pair > .syntax--punctuation {
  color: #abb2bf;
}
.syntax--source.syntax--ts .syntax--keyword.syntax--operator {
  color: #56b6c2;
}
.syntax--source.syntax--flow .syntax--keyword.syntax--operator {
  color: #56b6c2;
}
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--string.syntax--quoted.syntax--json {
  color: #e06c75;
}
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--string.syntax--quoted.syntax--json > .syntax--punctuation.syntax--string {
  color: #e06c75;
}
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--value.syntax--json > .syntax--string.syntax--quoted.syntax--json,
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--array.syntax--json > .syntax--value.syntax--json > .syntax--string.syntax--quoted.syntax--json,
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--value.syntax--json > .syntax--string.syntax--quoted.syntax--json > .syntax--punctuation,
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--array.syntax--json > .syntax--value.syntax--json > .syntax--string.syntax--quoted.syntax--json > .syntax--punctuation {
  color: #98c379;
}
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--dictionary.syntax--json > .syntax--constant.syntax--language.syntax--json,
.syntax--source.syntax--json .syntax--meta.syntax--structure.syntax--array.syntax--json > .syntax--constant.syntax--language.syntax--json {
  color: #56b6c2;
}
.syntax--ng.syntax--interpolation {
  color: #e06c75;
}
.syntax--ng.syntax--interpolation.syntax--begin,
.syntax--ng.syntax--interpolation.syntax--end {
  color: #61afef;
}
.syntax--ng.syntax--interpolation .syntax--function {
  color: #e06c75;
}
.syntax--ng.syntax--interpolation .syntax--function.syntax--begin,
.syntax--ng.syntax--interpolation .syntax--function.syntax--end {
  color: #61afef;
}
.syntax--ng.syntax--interpolation .syntax--bool {
  color: #d19a66;
}
.syntax--ng.syntax--interpolation .syntax--bracket {
  color: #abb2bf;
}
.syntax--ng.syntax--pipe,
.syntax--ng.syntax--operator {
  color: #abb2bf;
}
.syntax--ng.syntax--tag {
  color: #56b6c2;
}
.syntax--ng.syntax--attribute-with-value .syntax--attribute-name {
  color: #e5c07b;
}
.syntax--ng.syntax--attribute-with-value .syntax--string {
  color: #c678dd;
}
.syntax--ng.syntax--attribute-with-value .syntax--string.syntax--begin,
.syntax--ng.syntax--attribute-with-value .syntax--string.syntax--end {
  color: #abb2bf;
}
.syntax--source.syntax--php .syntax--class.syntax--bracket {
  color: #abb2bf;
}
/*
   This defines styling rules for syntax classes.

   See the naming conventions for a list of syntax classes:
   https://flight-manual.atom.io/hacking-atom/sections/syntax-naming-conventions

   When styling rules conflict:
   - The last rule overrides previous rules.
   - The rule with most classes and pseudo-classes overrides the last rule.
*/
.syntax--keyword {
  color: #c678dd;
}
.syntax--keyword.syntax--type {
  color: #56b6c2;
}
.syntax--keyword.syntax--function {
  color: #e06c75;
}
.syntax--keyword.syntax--variable {
  color: #e06c75;
}
.syntax--entity {
  color: #abb2bf;
}
.syntax--entity.syntax--parameter {
  color: #abb2bf;
}
.syntax--entity.syntax--support {
  color: #e06c75;
}
.syntax--entity.syntax--decorator:last-child {
  color: #61afef;
}
.syntax--entity.syntax--label {
  text-decoration: underline;
}
.syntax--entity.syntax--function {
  color: #61afef;
}
.syntax--entity.syntax--operator {
  color: #61afef;
}
.syntax--entity.syntax--operator.syntax--symbolic {
  color: #abb2bf;
}
.syntax--entity.syntax--type {
  color: #56b6c2;
}
.syntax--entity.syntax--tag {
  color: #e06c75;
}
.syntax--entity.syntax--attribute {
  color: #d19a66;
}
.syntax--punctuation {
  color: #abb2bf;
}
.syntax--punctuation.syntax--accessor {
  color: #abb2bf;
}
.syntax--punctuation.syntax--accessor.syntax--member,
.syntax--punctuation.syntax--accessor.syntax--scope {
  color: #c678dd;
}
.syntax--punctuation.syntax--embedded {
  color: #c678dd;
}
.syntax--string {
  color: #98c379;
}
.syntax--string.syntax--immutable {
  color: #98c379;
}
.syntax--string.syntax--part {
  color: #56b6c2;
}
.syntax--string.syntax--interpolation {
  color: #c678dd;
}
.syntax--string.syntax--regexp {
  color: #98c379;
}
.syntax--string.syntax--regexp.syntax--language {
  color: #c678dd;
}
.syntax--string.syntax--regexp.syntax--variable {
  color: #61afef;
}
.syntax--string.syntax--regexp.syntax--punctuation {
  color: #c678dd;
}
.syntax--constant {
  color: #d19a66;
}
.syntax--constant.syntax--character {
  color: #98c379;
}
.syntax--constant.syntax--character.syntax--escape {
  color: #98c379;
}
.syntax--constant.syntax--character.syntax--code {
  color: #56b6c2;
}
.syntax--text {
  color: #abb2bf;
}
.syntax--markup.syntax--heading {
  color: #e06c75;
}
.syntax--markup.syntax--list.syntax--punctuation {
  color: #e06c75;
}
.syntax--markup.syntax--quote {
  color: #5c6370;
  font-style: italic;
}
.syntax--markup.syntax--bold {
  color: #d19a66;
  font-weight: bold;
}
.syntax--markup.syntax--italic {
  color: #c678dd;
  font-style: italic;
}
.syntax--markup.syntax--underline {
  color: #56b6c2;
  text-decoration: underline;
}
.syntax--markup.syntax--strike {
  color: #e06c75;
}
.syntax--markup.syntax--raw {
  color: #98c379;
}
.syntax--markup.syntax--link {
  color: #56b6c2;
}
.syntax--markup.syntax--alt {
  color: #61afef;
}
.syntax--markup.syntax--inserted {
  color: #98c379;
}
.syntax--markup.syntax--inserted .syntax--punctuation {
  color: #98c379;
}
.syntax--markup.syntax--highlighted {
  color: #98c379;
}
.syntax--markup.syntax--highlighted .syntax--punctuation {
  color: #98c379;
}
.syntax--markup.syntax--deleted {
  color: #e06c75;
}
.syntax--markup.syntax--deleted .syntax--punctuation {
  color: #e06c75;
}
.syntax--markup.syntax--changed {
  color: #c678dd;
}
.syntax--markup.syntax--changed .syntax--punctuation {
  color: #c678dd;
}
.syntax--markup.syntax--commented {
  color: #5c6370;
}
.syntax--markup.syntax--commented .syntax--punctuation {
  color: #5c6370;
}
.syntax--comment {
  color: #5c6370;
  font-style: italic;
}
.syntax--comment.syntax--caption {
  color: #6a7181;
  font-weight: bold;
}
.syntax--comment.syntax--term {
  color: #707989;
}
.syntax--comment.syntax--punctuation {
  color: #5c6370;
  font-weight: normal;
}
.syntax--invalid:not(.syntax--punctuation).syntax--illegal {
  color: white !important;
  background-color: #e05252 !important;
}
.syntax--invalid:not(.syntax--punctuation).syntax--deprecated {
  color: #523d14 !important;
  background-color: #e0c285 !important;
}
.syntax--source.syntax--css .syntax--entity.syntax--function {
  color: #828997;
}
.syntax--source.syntax--css .syntax--entity.syntax--function.syntax--support {
  color: #56b6c2;
}
.syntax--source.syntax--css .syntax--entity.syntax--selector {
  color: #d19a66;
}
.syntax--source.syntax--css .syntax--entity.syntax--selector.syntax--tag {
  color: #e06c75;
}
.syntax--source.syntax--css .syntax--entity.syntax--selector.syntax--id {
  color: #61afef;
}
.syntax--source.syntax--css .syntax--entity.syntax--property {
  color: #828997;
}
.syntax--source.syntax--css .syntax--entity.syntax--property.syntax--support {
  color: #abb2bf;
}
.syntax--source.syntax--css .syntax--entity.syntax--variable {
  color: #e06c75;
}
.syntax--source.syntax--css .syntax--constant {
  color: #828997;
}
.syntax--source.syntax--css .syntax--constant.syntax--support {
  color: #abb2bf;
}
.syntax--source.syntax--css .syntax--constant.syntax--numeric {
  color: #d19a66;
}
.syntax--source.syntax--css .syntax--constant.syntax--media {
  color: #d19a66;
}
.syntax--source.syntax--css .syntax--constant.syntax--color {
  color: #d19a66;
}
.syntax--source.syntax--css .syntax--constant.syntax--offset {
  color: #abb2bf;
}
.syntax--source.syntax--css .syntax--constant.syntax--attribute-value {
  color: #98c379;
}
.syntax--source.syntax--css .syntax--punctuation.syntax--selector {
  color: #d19a66;
}
.syntax--source.syntax--css .syntax--punctuation.syntax--selector.syntax--wildcard {
  color: #e06c75;
}
.syntax--source.syntax--css .syntax--punctuation.syntax--selector.syntax--id {
  color: #61afef;
}
.syntax--source.syntax--css .syntax--punctuation.syntax--selector.syntax--attribute {
  color: #abb2bf;
}
</style>

  </head>
  <body>
    <p><strong>Table of Contents</strong></p>
<!-- TOC -->
<ul>
<li><a href="#download-mar-databases">Download MAR databases</a></li>
<li><a href="#install-programs">Install programs</a></li>
<li><a href="#prepare-databases">Prepare databases</a>
<ul>
<li><a href="#check-databases-for-duplicates">Check databases for duplicates</a></li>
<li><a href="#rename-sequences">Rename sequences:</a></li>
<li><a href="#translate-to-aa">Translate to aa</a></li>
</ul>
</li>
<li><a href="#using-hmms-to-fish-out-of-our-protein-nosz-from-reference-database">Using HMMs to fish out of our protein (NosZ) from reference database.</a></li>
<li><a href="#prepare-nosz-sequences-alignment-and-tree">Prepare nosZ sequences alignment and tree</a></li>
<li><a href="#environmental-sequences--cheatcheating">Environmental sequences -cheatcheating</a></li>
<li><a href="#phylogenetic-placement-of-environmental-sequences">Phylogenetic placement of environmental sequences</a>
<ul>
<li><a href="#alignment-of-environmental-sequences-with-papara">Alignment of environmental sequences with PaPaRa</a></li>
<li><a href="#phylogenetic-placement-of-environmental-sequences-with-epa-ng">Phylogenetic placement of environmental sequences with EPA-ng</a></li>
</ul>
</li>
</ul>
<!-- /TOC -->
<h3>Download MAR databases</h3>
<p>MAR databases website: <a href="https://mmp.sfb.uit.no/downloads/">https://mmp.sfb.uit.no/downloads/</a></p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ wget '<span class="syntax--markup syntax--underline syntax--link syntax--https syntax--hyperlink">https://public.sfb.uit.no/MarDB/BLAST/nucleotides/mardb_nucleotides_V6.fna</span>' &amp;
$ wget '<span class="syntax--markup syntax--underline syntax--link syntax--https syntax--hyperlink">https://public.sfb.uit.no/MarRef/BLAST/nucleotides/marref_nucleotides_V6.fna</span>' &amp;</span></span></pre>
<h3>Install programs</h3>
<p>Install miniconda
<a href="https://docs.conda.io/projects/conda/en/latest/user-guide/install/linux.html">https://docs.conda.io/projects/conda/en/latest/user-guide/install/linux.html</a></p>
<p>configure channels</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ conda config --add channels defaults
$ conda config --add channels bioconda
$ conda config --add channels conda-forge</span></span></pre>
<p>create a new environment</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ conda create -n traits
$ conda install python=3.9.6
$ conda install -c conda-forge biopython
$ conda install -c bioconda seqkit prodigal hmmer muscl iqtree epa-ng gappa</span></span></pre>
<p>Install Papara
<a href="https://cme.h-its.org/exelixis/web/software/papara/index.html">https://cme.h-its.org/exelixis/web/software/papara/index.html</a></p>
<h3>Prepare databases</h3>
<h4>Check databases for duplicates</h4>
<p><strong>MarRef database - do not run</strong></p>
<p>Removing duplicates based on names (-n) not on sequences, saving the number and list of duplicate seqs (-D) and duplicate seqs (-d)</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ cat marref_nucleotides_V6.fna | seqkit rmdup -n -o marrefnoDup_nucleotides.fna -d duplicates_marref.fna -D duplicates_marref_info.txt
[INFO] 0 duplicated records removed</span></span></pre>
<p><strong>MarDB database</strong></p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ cat mardb_nucleotides_V6.fna | seqkit rmdup -n -o mardbnoDup_nucleotides.fna -d duplicates_mardb.fna -D duplicates_mardb_info.txt 1&gt;log_rmdup.txt 2&gt;&amp;1 &amp;</span></span></pre>
<p><strong>prepare a single fasta</strong></p>
<p>Join the two databases in a single fasta:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ cat marref_nucleotides_V6.fna mardbnoDup_nucleotides.fna &gt; mar_nucleotides.fna &amp;</span></span></pre>
<h4>Rename sequences:</h4>
<p>Custom python script, first version:
<code style="font-family: Menlo, Consolas, &quot;DejaVu Sans Mono&quot;, monospace;">rename_mar_seqs.py</code></p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">from Bio.SeqIO.FastaIO import SimpleFastaParser
</span>
<span class="syntax--meta syntax--paragraph syntax--text">newseqs = {}
equivalent = {}
# parse fasta file with the low-level SimpleFastaParser, reads it as a tuple
with open("mar_nucleotides.fna") as sequences:
</span>	<span class="syntax--meta syntax--paragraph syntax--text">for k, seq in enumerate(SimpleFastaParser(sequences)):
</span>		<span class="syntax--meta syntax--paragraph syntax--text">newseqs[k]=seq[1]
		equivalent[k]=seq[0]
</span>

<span class="syntax--meta syntax--paragraph syntax--text">ofile = open("mar_renamed_nucleotides.fna", "w")
for i in newseqs.keys():
</span>	<span class="syntax--meta syntax--paragraph syntax--text">ofile.write("&gt;{}\n{}\n".format(i, newseqs[i]))
</span><span class="syntax--meta syntax--paragraph syntax--text">ofile.close()
</span>

<span class="syntax--meta syntax--paragraph syntax--text">ofile = open("mar_names_equivalence.txt", "w")
for i in equivalent.keys():
</span>	<span class="syntax--meta syntax--paragraph syntax--text">ofile.write("{}\t{}\n".format(i, equivalent[i]))
</span><span class="syntax--meta syntax--paragraph syntax--text">ofile.close()</span></span></pre>
<p>Copy to your folder where the fasta file is, check permisions with chmod and run it:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ chmod u+x rename_mar_seqs.py
$ python rename_mar_seqs.py 1&gt;log_rename.txt 2&gt;&amp;1 &amp;</span></span></pre>
<h4>Translate to aa</h4>
<p><strong>Using prodigal</strong> in anonimous mode (takes days):</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ prodigal -v
Prodigal V2.6.3: February, 2016
$ prodigal -i mar_renamed_nucleotides.fna -o mar_gene_coords.gbk -a mar_proteintrans.faa -p meta 2&gt;log_prodigal.err &amp;</span></span></pre>
<p>I have prepared a smaller dataset that contains the 10% of MarRef only as an example to run prodigal in 30-40 min. Copy marsample.fna file to your working directory.</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ prodigal -i marsample.fna -o marsample_gene_coords.gbk -a marsample_proteintrans.faa -p meta 2&gt;log_prodigal.err &amp;</span></span></pre>
<p>Clean fasta headers from prodigal output (small dataset):</p>
<p>Small dataset:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ gawk 'BEGUIN{FS="#"}{if ($1 ~ /^&gt;/) print $1; else print $0}' marsample_proteintrans.faa |\
</span> <span class="syntax--meta syntax--paragraph syntax--text">sed 's/_1//g' &gt; marsample_proteintrans_clean.faa</span></span></pre>
<p>Mar DBs:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ gawk 'BEGUIN{FS="#"}{if ($1 ~ /^&gt;/) print $1; else print $0}' mar_proteintrans.faa |\
</span> <span class="syntax--meta syntax--paragraph syntax--text">sed 's/_1//g' &gt; mar_proteintrans_clean.faa 2&gt;err_gawk.txt &amp;</span></span></pre>
<p><strong>Alternative without gene prediction:</strong></p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">seqkit translate -T 11 --trim mar_renamed_nucleotides.fna &gt; mar_proteintrans_v2.faa 2&gt;log_seqkit.txt &amp;</span></span></pre>
<p>No need to clean sequence headers</p>
<h3>Using HMMs to fish out of our protein (NosZ) from reference database.</h3>
<p>Download hmm files:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ wget '<span class="syntax--markup syntax--underline syntax--link syntax--https syntax--hyperlink">https://ftp.ncbi.nlm.nih.gov/hmm/current/hmm_PGAP.HMM/TIGR04246.1.HMM</span>'
$ wget '<span class="syntax--markup syntax--underline syntax--link syntax--https syntax--hyperlink">https://ftp.ncbi.nlm.nih.gov/hmm/current/hmm_PGAP.HMM/TIGR04244.1.HMM</span>'</span></span></pre>
<p>To use both HMM, to put them toguether in a single file.</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ cat TIGR04244.1.HMM TIGR04246.1.HMM &gt; TIGRnosZ.HMM</span></span></pre>
<p>Now, we can use HMMER to search for nosZ in the database. We are using the gathering score from the models. With the sample database:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">hmmsearch --noali --domtblout marsample_nosZ_tigrfam.hmm --cut_ga --cpu 15 TIGRnosZ.HMM marsample_proteintrans_clean.faa &gt; /dev/null 2&gt;hmmerrlog.txt &amp;</span></span></pre>
<p>Full database:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">hmmsearch --noali --domtblout /mnt/nfelbrus2/mar_db/mar_nosZ_tigrfam.hmm --cut_ga --cpu 15 /media/disk5/nfernandez/nosZ/TIGRnosZ/TIGRnosZ.HMM /mnt/nfelbrus2/mar_db/mar_proteintrans_clean.faa &gt; /dev/null 2&gt;/mnt/nfelbrus2/mar_db/hmmerrlog.txt &amp;</span></span></pre>
<p><strong>Filter nosZ database to retrieve hits’ sequences:</strong></p>
<p>Once it is done, pick up sequence IDs from hmm output table with the python script.
This script produces a text file with hit´s IDs</p>
<p><code style="font-family: Menlo, Consolas, &quot;DejaVu Sans Mono&quot;, monospace;">parse_marsampleHMM.py</code></p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">from Bio import SearchIO
import csv
</span>
<span class="syntax--meta syntax--paragraph syntax--text"># Read the output table of database search with TIGRFAMs HMMs for nosZ
hmm_qresult = SearchIO.parse('marsample_nosZ_tigrfam.hmm',  'hmmsearch3-domtab')
</span>
<span class="syntax--meta syntax--paragraph syntax--text"># Filter table to make a list of hits IDs
hit_ids = []
for qresult in hmm_qresult:
</span>    <span class="syntax--meta syntax--paragraph syntax--text">for i in range(len(qresult)):
</span>        <span class="syntax--meta syntax--paragraph syntax--text">hit_ids.append(qresult[i].id)
</span>
<span class="syntax--meta syntax--paragraph syntax--text"># Write list to a text file
file = open('hits_nosZ_marsample.txt', 'w')
for index in range(len(hit_ids)):
</span>    <span class="syntax--meta syntax--paragraph syntax--text">file.write(str(hit_ids[index]) + "\n")
</span><span class="syntax--meta syntax--paragraph syntax--text">file.close()</span></span></pre>
<p>Let´s run it and have a look:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ python parse_marsampleHMM.py
$ less hits_nosZ_marsample.txt</span></span></pre>
<p>Now, <strong>filter database</strong> using seqkit tools to keep sequences included in <code style="font-family: Menlo, Consolas, &quot;DejaVu Sans Mono&quot;, monospace;">hits_nosZ_marsample.txt</code></p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ seqkit grep -f hits_nosZ_marsample.txt marsample_proteintrans_clean.faa -o hits_nosZ_marsample.faa</span></span></pre>
<p>This step produces a fasta file, <code style="font-family: Menlo, Consolas, &quot;DejaVu Sans Mono&quot;, monospace;">hits_nosZ_marsample.faa</code>, which contains the putative nosZ sequences only.</p>
<p>Commands to <strong>work with the full database</strong> are similar, just change names/paths also in the python script:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ python parse_marNosZhmm.py
$ wc -l hits_mar_nosZ.txt&nbsp;
1469 hits_mar_nosZ.txt</span></span></pre>
<p>The <code style="font-family: Menlo, Consolas, &quot;DejaVu Sans Mono&quot;, monospace;">parse_marNosZhmm.py</code></p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">from Bio import SearchIO
import csv
</span>
<span class="syntax--meta syntax--paragraph syntax--text"># Read the output table of MarDB search with TIGRFAMs HMMs for nosZ
hmm_qresult = SearchIO.parse('mar_nosZ_tigrfam.hmm',  'hmmsearch3-domtab')
</span>
<span class="syntax--meta syntax--paragraph syntax--text"># Filter table to make a list of hits IDs
hit_ids = []
for qresult in hmm_qresult:
</span>    <span class="syntax--meta syntax--paragraph syntax--text">for i in range(len(qresult)):
</span>        <span class="syntax--meta syntax--paragraph syntax--text">hit_ids.append(qresult[i].id)
</span>
<span class="syntax--meta syntax--paragraph syntax--text"># Write list to a text file
file = open('hits_mar_nosZ.txt', 'w')
for index in range(len(hit_ids)):
</span>    <span class="syntax--meta syntax--paragraph syntax--text">file.write(str(hit_ids[index]) + "\n")
</span><span class="syntax--meta syntax--paragraph syntax--text">file.close()</span></span></pre>
<p>and get the nosZ sequences:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ seqkit grep -f hits_mar_nosZ.txt mar_proteintrans_clean.faa -o mar_nosZ.faa 1&gt;/dev/null 2&gt;&amp;1 &amp;</span></span></pre>
<h3>Prepare nosZ sequences alignment and tree</h3>
<p>Now, align nosZ squences:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ mkdir alignments
$ muscle -in hits_nosZ_mar.faa -phyiout alignments/hits_nosZ_aligned.phy -fastaout alignments/hits_nosZ_aligned.faa</span></span></pre>
<p>Build a tree with iqTREE. Including an initial step to find the best-fid model:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ iqtree -nt AUTO -ntmax 15 -s alignments/hits_nosZ_aligned.faa</span></span></pre>
<p>With the full database (changed directory):</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ mkdir alignments
$ muscle -in mar_nosZ.faa -phyiout alignments/mar_nosZ_aligned.phy -fastaout alignments/mar_nosZ_aligned.faa 1&gt;muscle_log.txt 2&gt;&amp;1 &amp;</span></span></pre>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ iqtree -nt AUTO -ntmax 18 -s alignments/mar_nosZ_aligned.phy 1&gt;log_iqtree.txt 2&gt;&amp;1 &amp;</span></span></pre>
<p>[1] 15498 - Still running when finishing this document…</p>
<h3>Environmental sequences -cheatcheating</h3>
<p><strong>Fungene sequences</strong></p>
<p><a href="http://fungene.cme.msu.edu/">http://fungene.cme.msu.edu/</a></p>
<p>Downloaded protein sequences on 26-Aug-2021 the three datasets available for nosZ:
nosZ-Blaz Stres
nosZ_atypical_1-Roberto Sandford
nosZ_atypical_2-Roberto Sandford</p>
<p>Join three datasets with cat and remove duplicates:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ cat fungene_nosZ.faa | seqkit rmdup -n -o fungene_nosZ_noDup.faa -d duplicated_fungene_nosZ.faa -D duplicated_fungene_nosZ_detail.txt
[INFO] 12926 duplicated records removed</span></span></pre>
<p>Avoid long names. Remove everything but the nucleotide GI that is the first field</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ gawk 'BEGIN{FS=" "}{if ($1 ~ /^&gt;/) print $1; else print $0 }' fungene_nosZ_noDup.faa &gt; fungeneTest.faa
$ wc -l fungeneTest.faa
89856 fungeneTest.faa</span></span></pre>
<p>Prepare a small subsample of Fungene sequences. Notice that fungene sequences are lowercase, but mardbs are in uppercase. Fix it with seqkit</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ head -n 55 fungeneTest.faa | seqkit seq -u &gt; query.faa</span></span></pre>
<h3>Phylogenetic placement of environmental sequences</h3>
<h4>Alignment of environmental sequences with PaPaRa</h4>
<p><strong>Run PaPaRa</strong> with the -r option to ensure no additional gaps are included in the original MSA. Other options are -a for aminoacid sequences, -t reference tree, -s reference alignment -q sequences to align (query sequences), -j threads, -n sufix for output:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ cd alignments/
$ papara -t hits_nosZ_aligned.phy.treefile -a -r -s hits_nosZ_aligned.phy -q ../query.faa -j 4 -n papout 1&gt;log_papara.txt 2&gt;&amp;1 &amp;</span></span></pre>
<p>PaPaRa outputs a MSA including both the reference and query sequences, but EPA-ng
needs them separated, it is easy to separate:</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ epa-ng --split hits_nosZ_aligned.phy papara_alignment.papout
INFO Splitting files based on reference: hits_nosZ_aligned.phy</span></span></pre>
<p>it produces two files: reference.fasta and query.fasta which have the aligned sets.</p>
<h4>Phylogenetic placement of environmental sequences with EPA-ng</h4>
<p><strong>Pace the query sequences in the reference tree</strong> with EPA-ng:</p>
<p>Make output directory first and use the MSAs in fasta format produced by epa-ng --split.</p>
<p>We have to indicate the substitution model used in the three.</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ mkdir epa-ng-test
$ epa-ng --tree hits_nosZ_aligned.phy.treefile --ref-msa reference.fasta --query query.fasta --out-dir epa-ng-test --model LG+F+R10 &gt; log_epang.txt 2&gt;&amp;1 &amp;</span></span></pre>
<p>Now, we can <strong>start analyzing the results</strong> with gappa</p>
<pre class="editor-colors lang-text" style="tab-size: 2;"><span class="syntax--text syntax--plain"><span class="syntax--meta syntax--paragraph syntax--text">$ gappa examine graft --jplace-path epa_result.jplace --name-prefix FUNGEN --out-dir gappaout</span></span></pre>

  </body>
</html>
